packages/services/src/kernel/manager.ts

```diff
@@ -195,71 +195,126 @@ export class KernelManager implements Kernel.IManager {
    * This will emit [[runningChanged]] if the running kernels list
    * changes.
    */
-  shutdown(id: string): Promise<void> {
-    let index = ArrayExt.findFirstIndex(this._models, value => value.id === id);
+  async shutdown(id: string): Promise<void> {
+    const models = this._models;
+    const kernels = this._kernels;
+    const index = ArrayExt.findFirstIndex(models, value => value.id === id);
+
     if (index === -1) {
       return;
     }
 
     // Proactively remove the model.
-    this._models.splice(index, 1);
-    this._runningChanged.emit(this._models.slice());
-
-    return Kernel.shutdown(id, this.serverSettings).then(() => {
-      let toRemove: Kernel.IKernel[] = [];
-      this._kernels.forEach(k => {
-        if (k.id === id) {
-          k.dispose();
-          toRemove.push(k);
-        }
-      });
-      toRemove.forEach(k => {
-        this._kernels.delete(k);
-      });
+    models.splice(index, 1);
+
+    // Delete and dispose the kernel locally.
+    kernels.forEach(kernel => {
+      if (kernel.id === id) {
+        kernels.delete(kernel);
+        kernel.dispose();
+      }
     });
+
+    // Shut down the remote session.
+    await Kernel.shutdown(id, this.serverSettings);
+
+    // Emit the new model list.
+    this._runningChanged.emit(models.slice());
   }
 
   /**
    * Shut down all kernels.
    *
    * @returns A promise that resolves when all of the kernels are shut down.
    */
-  shutdownAll(): Promise<void> {
-    // Proactively remove all models.
-    let models = this._models;
-    if (models.length > 0) {
-      this._models = [];
-      this._runningChanged.emit([]);
+  async shutdownAll(): Promise<void> {
+    // Remove all models.
+    if (this._models.length) {
+      this._models.length = 0;
     }
 
-    return this._refreshRunning().then(() => {
-      return Promise.all(
-        models.map(model => {
-          return Kernel.shutdown(model.id, this.serverSettings).then(() => {
-            let toRemove: Kernel.IKernel[] = [];
-            this._kernels.forEach(k => {
-              k.dispose();
-              toRemove.push(k);
-            });
-            toRemove.forEach(k => {
-              this._kernels.delete(k);
-            });
-          });
-        })
-      ).then(() => {
-        return undefined;
-      });
+    // Emit the new model list without waiting for API requests.
+    this._runningChanged.emit([]);
```

Comment:
We should only emit if we actually did truncate, so move this back up into the if check above.

Reply:
(I'm going ahead and doing some of this that I can put in as a PR against this PR.)

Reply:
Resolved in https://github.com/afshin/jupyterlab/pull/24